### PR TITLE
Sequence decoding submodule

### PR DESCRIPTION
**Pull request by bmcfee**
#### Reference Issue
Implements #639 

#### What does this implement/fix? Explain your changes.

This PR adds a numba-accelerated implementation of the Viterbi algorithm for state sequence decoding.

It's in progress, but the target functionality is:

- [x] `viterbi`: standard Viterbi decoding for observation likelihood sequences `P[X(t) | y(t)]`
- [x] `viterbi_d`: Viterbi decoding for conditional state estimates `P[Y(t) | X(t)]`
- [x] `viterbi_ml`: Viterbi decoding for multi-label conditional state estimates.  (Reduces to a collection of two-state `viterbi_d` problems)
- [x] Transition matrix constructors
  - [x] uniform
  - [x] self-loop
  - [x] cycle
  - [x] local
- [x] Full unit tests

---
This change is [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)

### CODE REVIEW

**Comment by stefan-balke** [comment]
Hey @bmcfee,

that's an awesome start. Could we have some small example? Maybe an F0-trajectory or even easier?
Would make it much easier from the outside to provide some feedback!

Do you use the notation from your chapter in "Computational Sound Scenes" or could you give a literature pointer? Could be beneficial for users to understand the code with an accompanying website/textbook. What you think?

**Comment by bmcfee** [comment]
> Could we have some small example? Maybe an F0-trajectory or even easier?
> Would make it much easier from the outside to provide some feedback!

It's a little tricky to make a small, self-contained example that's still motivated by audio.  The reason being that `viterbi()` expects input of the form `prob[t, s] = P[Obs(t) | State(t) = s]` (observation likelihood), but we typically have the reverse situation in audio: `P[State(t) = s | Obs(t)]` (i.e., predict the class from the observed features at time t).  The latter case will be handled by `viterbi_d`.

That said, the unit tests include a small example taken from the wikipedia article on Viterbi decoding, so it might be a good place to get started.



> Do you use the notation from your chapter in "Computational Sound Scenes" or could you give a literature pointer? 

Not exactly.  Much of the notation in CASSE was decided by the editors for consistency across chapters (which makes sense).  I'd prefer here to use notation that's internally consistent to librosa.

Otherwise, there are tons of good references for people to learn about Viterbi.  My default would be Russell & Norvig's textbook, but maybe it would be better to link out to something that's open access?

**Comment by bmcfee** [comment]
`viterbi_d` (discriminative decoding) is done now, but could still use a non-trivial (but simple) unit test.

**Comment by bmcfee** [comment]
Quick update: I'll probably add an example notebook using `viterbi_d` to denoise predictions for silence detection.  We'll need to dig up an appropriate audio example, but it should be pretty straightforward.

**Comment by bmcfee** [comment]
Expanding a bit on the description in the PR, here's a more detailed discussion of how each of the functions in the sequence module works.

## Decoders

- `viterbi`: this is the standard Viterbi algorithm.  It is tested against a worked example provided on Wikipedia, and matches state sequence likelihood to numerical tolerance.  It takes as input:
  - the observation likelihoods `prob = P[X(t) | Y(t)]`
  - the transition matrix `transition = P[Y(t+1) | Y(t)]`, and 
  - the initial state distribution `p_init = P[Y(0)]`.

- `viterbi_d`: this is intended for use with discriminative "state" (multi-class) predictors.  It takes as input:
  - the conditional state likelihoods `prob = P[Y(t) | X(t)]`
  - the transition matrix `transition = P[Y(t+1) | Y(t)]`
  - the marginal state distribution `p_state = P[Y]`
  - the initial state distribution `p_init = P[Y(0)]` (assumed to = `p_state` if not provided)

  - As outlined in #639, the idea here is to use Bayes rule to convert the state likelihoods into unnormalized observation likelihoods:
    ```
    P[X(t) | Y(t)] = P[X(t)] * P[Y(t) | X(t)] / P[Y(t)]
                   ∝ P[Y(t) | X(t)] / P[Y(t)]
    ```
    Since `P[X(t)]` is constant for the purposes of conditional inference, we ignore this factor.  The denominator `P[Y(t)]` need not generally be equal to the initial state distribution, so we allow the user to specify it separately.  The result is proportional to the observation likelihood, and we can use `viterbi` to compute the maximum likelihood path.  (We don't actually use `viterbi`, but instead wrap directly to the accelerated `_viterbi` helper function.)

- `viterbi_ml`: this is a convenience function intended for multi-label sequence problems (e.g., source identification).  Since the labels are not mutually exclusive, we decode each one separately as a binary sequence problem.  The inputs are
  - the conditional label likelihoods `prob[s, t] = P[Y_s(t) = 1 | X(t)]`
  - 2-by-2 transition matrices for each label `transition[s, i, j] = P[Y_s(t+1) = i | Y_s(t) = j]`
  - marginal state distributions `p_state[s] = P[Y_s = 1]`
  - initial state distributions `p_init[s] = P[Y_s(0) = 1]`

  - The output is a binary matrix `states[s, t]` indicating whether state `s` is active at time `t`.
  - Each state `s` is independently decoded by inferring the parameters of the corresponding 2-state problem, and calling out to `viterbi_d`.

## Transition builders

- `transition_uniform`: make an `n_states * n_states` matrix of values `1./n_states`.  This isn't super useful on its own, but can be handy if mixed in as a regularizer to a structured transition matrix.

- `transition_loop`: self-loop with probability `p`, and the remaining `1-p` mass is spread uniformly over the other `n-1` states.  If `p` is a vector, `p[s]` gives the self-loop probability for state `s`.  This is useful for unstructured state spaces with temporal locality.

- `transition_cycle`: self-loop with probability `p`, and `transition[i, i+1] = 1-p`.  `p` can be vectorized as above. This is useful for cyclically structured state spaces (eg metrical position).

- `transition_local`: use a user-specified window function so that `transition[i, j] ∝ window(i - j)`.  The user specifies the width of the window `w` as well, and `|i-j|>w => transition[i, j] = 0`.  `w` can be vectorized.  Optionally, the window can wrap around the transition matrix, giving a softened and bidirectional alternative to `transition_cycle`.  This builder is useful for state spaces with local affinity, such as discretized frequency bands used in fundamental frequency estimation.  (pYin, for example, uses triangular windows.)

**Comment by bmcfee** [comment]
@stefan-balke here's a link to the [advanced example notebook](https://gist.github.com/bmcfee/bb5f7f34ff4a8d9a8028791c5d3ddb92) using viterbi_d for silence detection.

**Comment by stefan-balke** [comment]
Hey @bmcfee that's exactly what I mean. Will follow-up your code in the next days (have some long train trips :))

**Comment by bmcfee** [comment]
@stefan-balke thanks!

Otherwise, I think this one's ready for CR.

I added some unit tests for `viterbi_d` and `viterbi_ml`.  I'm not 100% happy with them, but couldn't come up with anything better.  Brief description:

- `viterbi_d` tests on a simple problem I cooked up and solved by brute-force enumeration.  It only checks the solution path (not the cost).
- `viterbi_ml` is an integration test: it checks that `viterbi_ml` and `viterbi_d` agree on a simple test problem.

**Comment by bmcfee** [comment]
After playing with unit tests for a while, I suspect there's a thinko in the discriminative->generative reduction.  (TLDR: The Bayes rule transformation should operate over label sequences, but instead assumes a static distribution, which is not correct.)  This will take a bit more work to get right.

**Comment by bmcfee** [comment]
updates: `viterbi_d` is now `viterbi_discriminative` and `viterbi_ml` is now `viterbi_binary`.

I think that does it here.

**Comment by bmcfee** [comment]
Side note: I'd like to get this merged in the next couple of days so we can ship 0.6.1 before the end of the month.  @stefan-balke do you have time for a quick CR?

**Comment by stefan-balke** [comment]
Yes.
First thing: why is it not located in core/?
Also DTW is somehow related to „sequences“. Should we put it together somewhere?

**Comment by bmcfee** [comment]
> First thing: why is it not located in core/?

`core` is getting bloated, and this is a pretty self-contained set of functions.

> Also DTW is somehow related to „sequences“. Should we put it together somewhere?

That's a good point.  I'd be okay with migrating `dtw` into the `sequence` module, but that's a slightly more delicate operation, since things like `librosa.core.dtw` and `librosa.dtw` would stop working.  We have do scaffolding in place for function migration that would throw a deprecation warning, so it's totally possible.  Would you be ok with it moving to `librosa.sequence.dtw`?

**Comment by stefan-balke** [comment]
Yeah. Moving DTW is fine for me.

**Comment by bmcfee** [comment]
Ok.  I'll probably move `fill_off_diagonal` to `util` in that case, and rename some of the (private) helper functions after moving into `sequence`.

**Comment by bmcfee** [comment]
Note: i'd like to do the dtw move as a separate pr, just to keep the CR simple.

**Comment by bmcfee** [comment]
last commit is a rebase to master and coverage fix, no functionality change.

**Comment by bmcfee** [comment]
Did a bit of squashing on this PR.  @stefan-balke anything else need attention?

**Comment by stefan-balke** [comment]
*[librosa/sequence.py, line 20 at r1](https://reviewable.io/reviews/librosa/librosa/710#-LD2YaPP1X05X3jwvSDS:-LD2YaPP1X05X3jwvSDT:b-omqnpz) ([raw file](https://github.com/librosa/librosa/blob/2ba9c7f749f574bfae378d4fe659acf5b830d7a4/librosa/sequence.py#L20)):*
> ```Python
>     Parameters
>     ----------
>     log_prob : np.ndarray, shape=(T, m)
> ```

Adapt docstrings to "librosa standard".

---


*Comments from [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)*

**Comment by stefan-balke** [comment]
*[librosa/sequence.py, line 158 at r1](https://reviewable.io/reviews/librosa/librosa/710#-LD2Yr716S01yqSnHpdb:-LD2Yr716S01yqSnHpdc:b-rtagwg) ([raw file](https://github.com/librosa/librosa/blob/2ba9c7f749f574bfae378d4fe659acf5b830d7a4/librosa/sequence.py#L158)):*
> ```Python
> 
> 
> # TODO
> ```

delete

---


*Comments from [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)*

**Comment by stefan-balke** [comment]
Hm...Reviewable is not on the latest commit...

File: librosa/sequence.py [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+def transition_local(n_states, width, window='triangle', wrap=False):
+    '''Construct a localized transition matrix.
+
+    The transition matrix will have the following properties:
+
+        - `transition[i, j] = 0` if `|i - j| > width`
+        - `transition[i, i]` is maximal
+        - `transition[i, i - width//2 : i + width//2]` has shape `window`
+
+    This type of transition matrix is appropriate for state spaces
```
Comment (stefan-balke) [inline librosa/sequence.py]: Can we have such a sentence for `viterbi()` and `viterbi_disrciminative()` just mentioning an MIR problem saying what are the observations (or states) and briefly explain what the function will then calculate.

**Comment by bmcfee** [comment]
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[librosa/sequence.py, line 54 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-54:-LD2kCgh2RMAuTvMGVyt:b1za61a) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L54)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

What is the difference between `prob` and `P`. If they are they same, they should be named the same.
</blockquote></details>

`prob` is used specifically to denote the emission likelihood, `P` is a generic probability notation

---

*[librosa/sequence.py, line 281 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-281:-LD2ldia5kwB-KPmgXus:biah5ie) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L281)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

`X` and `Y` were not introduced. Can you adapt this to the state probs etc?
</blockquote></details>

it's true for any rv's x and y, but i'll make a more explicit note

---


*Comments from [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)*

**Comment by bmcfee** [comment]
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[librosa/sequence.py, line 20 at r1](https://reviewable.io/reviews/librosa/librosa/710#-LD2YaPP1X05X3jwvSDS:-LD2mlk2C9BnSRp3VYM0:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/2ba9c7f749f574bfae378d4fe659acf5b830d7a4/librosa/sequence.py#L20)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

Adapt docstrings to "librosa standard".
</blockquote></details>

Done.

---

*[librosa/sequence.py, line 48 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-48:-LD2mn3u2m-LQ4lbm1KR:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L48)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

Docstring needs some adaptation to librosa "standard".
</blockquote></details>

Done.

---

*[librosa/sequence.py, line 115 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-115:-LD2mq2I6fP5sTuVFPpQ:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L115)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

Point to literature (Wikipedia + Russell & Norvig)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)*

**Comment by stefan-balke** [comment]
*[librosa/sequence.py, line 158 at r1](https://reviewable.io/reviews/librosa/librosa/710#-LD2Yr716S01yqSnHpdb:-LD2zBZU4ccjbdf47pxJ:blswxp1) ([raw file](https://github.com/librosa/librosa/blob/2ba9c7f749f574bfae378d4fe659acf5b830d7a4/librosa/sequence.py#L158)):*
<details><summary><i>Previously, bmcfee (Brian McFee) wrote…</i></summary><blockquote>

delete what?
</blockquote></details>

It's fine, this was a TODO list in an old version...

---


*Comments from [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)*

**Comment by bmcfee** [comment]
Review status: 0 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[librosa/sequence.py, line 281 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-281:-LD2wUwz7tiv9WToQ6sX:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L281)):*
<details><summary><i>Previously, bmcfee (Brian McFee) wrote…</i></summary><blockquote>

it's true for any rv's x and y, but i'll make a more explicit note
</blockquote></details>

Done.

---

*[librosa/sequence.py, line 353 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-353:-LD3GX-YCQqLZm6BC1c7:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L353)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

Example for the docs is missing.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)*

**Comment by bmcfee** [comment]
Review status: 0 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[librosa/sequence.py, line 158 at r1](https://reviewable.io/reviews/librosa/librosa/710#-LD2Yr716S01yqSnHpdb:-LD3GgTGAZdIgw9GcaBV:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/2ba9c7f749f574bfae378d4fe659acf5b830d7a4/librosa/sequence.py#L158)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

It's fine, this was a TODO list in an old version...
</blockquote></details>

Done.

---

*[librosa/sequence.py, line 144 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-144:-LD3GlBW6gE_lE65Hgmn:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L144)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

Example for the docs is missing.
</blockquote></details>

Done.

---

*[librosa/sequence.py, line 238 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-238:-LD3GhXy4o9_jnqmVpdc:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L238)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

Example for the docs is missing.
</blockquote></details>

Done.

---

*[librosa/sequence.py, line 600 at r5](https://reviewable.io/reviews/librosa/librosa/710#-LCB0zQHut9xLdH6jYFB-r5-600:-LD3GlvKBnC_PGW2TeE1:b-896fix) ([raw file](https://github.com/librosa/librosa/blob/643cf459e3118f8f3dd051e38055f40bd983a274/librosa/sequence.py#L600)):*
<details><summary><i>Previously, stefan-balke (Stefan Balke) wrote…</i></summary><blockquote>

Can we have such a sentence for `viterbi()` and `viterbi_disrciminative()` just mentioning an MIR problem saying what are the observations (or states) and briefly explain what the function will then calculate.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)*

**Comment by bmcfee** [comment]
Ok, I maybe went a little overboard, but I hacked up a quick template-based chord model for the `viterbi_discriminative` example.  Output figures look like
image

image


The other examples `viterbi` and `viterbi_binary` are text-based and self-contained.  I think this just about does it?

**Comment by lostanlen** [comment]
great disclaimer. 👍

**Comment by stefan-balke** [comment]
Yeah, thanks especially for the examples!

LGTM and thanks again @bmcfee!

**Comment by bmcfee** [comment]
Reviewed 1 of 3 files at r1, 3 of 6 files at r2, 2 of 3 files at r5, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/librosa/librosa/710)*

**Pull request merged**